connaisseur/flask_server.py

```diff
@@ -130,35 +129,34 @@ def __create_logging_msg(msg: str, **kwargs):
     return str({"message": msg, "context": dict(**kwargs)})
 
 
-def __create_json_patch(path: str, index: int, image: Image):
-    return {"op": "replace", "path": path.format(str(index)), "value": str(image)}
-
-
 def __admit(admission_request: AdmissionRequest):
-    policy = ImagePolicy()
-    parent_images = admission_request.wl_object.parent_images
-    patches = []
+    try:
```

Comment:
That `try` block is *very* long. Can we narrow it down to closer where the relevant exception could actually occur?

Reply:
This is intended. Whatever error may occur, we wanna catch it here and add some needed logging information to the error object, which we wouldn't have durring the initial raising of the error. This is then catched by yet another `try except` where the actual logging happens.

Reply:
That argument does not convince me, because you catch `BaseConnaisseurException` exceptions only, but the most part of the code does not raise that. -- Worse, `logging_context` depends on the loop iteration, but the `return` statement *after* the loop is also part of the `try` block. That means if the `return` statement somehow raises a `BaseConnaisseurException` exception, you will have a misleading logging context.

I looked which parts can throw that exception, and it seems like it's only the methods which are called on the policy and the validator. I would thus suggest to rewrite this as follows:

```python
def __admit(admission_request: AdmissionRequest):
    logging_context = dict(admission_request.context)
    policy = ImagePolicy()
    patches = []

    for type_index, image in admission_request.wl_object.containers.items():
        original_image = str(image)
        type_, index = type_index
        logging_context.update(image=original_image)

        # child resources have mutated image names, as their parents got mutated
        # before their creation. this may result in mismatch of rules or duplicate
        # lookups for already approved images. so child resources are automatically
        # approved without further check ups, when their parents were approved
        # earlier.
        if image in admission_request.wl_object.parent_containers.values():
            msg = f'automatic child approval for "{original_image}".'
            logging.info(__create_logging_msg(msg, **logging_context))
            continue

        try:
            policy_rule = policy.get_matching_rule(image)
            validator = CONFIG.get_validator(policy_rule.validator)

            msg = (
                f'starting verification of image "{original_image}" using rule '
                f'"{str(policy_rule)}" with arguments {str(policy_rule.arguments)}'
                f' and validator "{str(validator)}".'
            )
            logging.debug(
                __create_logging_msg(
                    msg, **logging_context, policy_rule=policy_rule, validator=validator
                )
            )

            trusted_digest = validator.validate(image, **policy_rule.arguments)
        except BaseConnaisseurException as err:
            err.update_context(**logging_context)
            raise err

        if trusted_digest:
            image.set_digest(trusted_digest)
            patches.append(admission_request.wl_object.get_json_patch(image, type_, index))
        msg = f'successful verification of image "{original_image}"'
        logging.info(__create_logging_msg(msg, **logging_context))
        return get_admission_review(admission_request.uid, True, patch=patches)
```

(Note that I also exchanged `+=` with `.append()`.)

Reply:
done.